src/smt/theory_str_noodler/nielsen_decision_procedure.h

```diff
@@ -0,0 +1,353 @@
+                for(const auto& tgt_symb : pr.second) {
+                    ret.add_edge(tgt_symb.first, pr.first, tgt_symb.second);
+                }
+            }
+            return ret;
+        }
+
+        /**
+         * @brief Remove redundant nodes (that cannot reach a final node)
+         * 
+         * @return NielsenGraph Nielsen graph without redundant nodes
+         */
+        TransitionGraph trim() const {
+            TransitionGraph rev = this->reverse();
+            TransitionGraph ret;
+            std::set<Formula> generated;
+            std::deque<Formula> worklist(this->fins.begin(), this->fins.end());
+            ret.set_init(this->init);
+            for(const Formula& fin : this->get_fins()) {
+                ret.add_fin(fin);
+            }
+            
+            while(!worklist.empty()) {
+                Formula nd = worklist.front();
+                generated.insert(nd);
+                worklist.pop_front();
+
+                for(const auto& tgt_symb : rev.edges[nd]) {
+                    ret.add_edge(tgt_symb.first, nd, tgt_symb.second);
+                    if(generated.find(tgt_symb.first) == generated.end()) {
+                        worklist.push_back(tgt_symb.first);
+                    }
+                }
+            }
+
+            return ret;
+        }
+
+        /**
+         * @brief Get shortest path from @p start to @p end.
+         * 
+         * @param start First node
+         * @param end Second node
+         * @return Path<Label> Visited nodes and labels on the shortest path.
+         */
+        Path<Label> shortest_path(const Formula& start, const Formula& end) const {
+            std::set<Formula> visited;
+            std::deque<std::pair<Formula, Path<Label>>> worklist;
+            worklist.push_back({start, {{start},{}}});
+
+            while(!worklist.empty()) {
+                auto elem = worklist.front();
+                worklist.pop_front();
+                if(visited.find(elem.first) != visited.end()) {
+                    continue;
+                }
+                visited.insert(elem.first);
+                if(elem.first == end) {
+                    return elem.second;
+                }
+
+                auto it = this->edges.find(elem.first);
+                if(it == this->edges.end()) continue;
+                for(const auto& tgt_symb : it->second) {
+                    if(visited.find(tgt_symb.first) == visited.end()) {
+                        std::vector<Label> path(elem.second.labels.begin(), elem.second.labels.end());
+                        path.push_back(tgt_symb.second);
+                        std::vector<Formula> nodes(elem.second.nodes.begin(), elem.second.nodes.end());
+                        nodes.push_back(tgt_symb.first);
+                        worklist.push_back({tgt_symb.first, {nodes, path}});
+                    }
+                }
+            }
+            util::throw_error("no shortest path");
```

Comment:
Could we maybe want to use this information? If so, we could return something like `std::optional<Path<Label>>` and return `std::nullopt` here which we can later check against instead of throwing an exception.

Reply:
Good idea, much clearer solution

Reply:
Well yeah, but if we never intend to use this information, it is unnecessary. This is not an interface function that we would expect anyone else than us to use. Therefore, we do not have to concern ourselves with whether the function is written in the most intuitive and clean way possible. If we are sure we will never want to compute something if there is no shortest path, then changing this is unnecessary and just a bit less efficient (because we have to unwrap the result in the `Some <value>` case).